src/Microsoft.Extensions.WebEncoders.Core/UrlPathDecoder.cs

```diff
@@ -0,0 +1,341 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+
+namespace Microsoft.Extensions.WebEncoders
+{
+    /// <summary>
+    /// UrlPathDecoder unescape a given path string.
+    /// </summary>
+    public class UrlPathDecoder
+    {
+        private static Encoding _utf8 = Encoding.GetEncoding(
+            "utf-8",
+            new EncoderExceptionFallback(),
+            new DecoderExceptionFallback());
+
+        /// <summary>
+        /// Unescape a url char array in place. Returns the length of the result.
+        /// 
+        /// - Everything is unescaped except %2F ('\')
+        /// - UTF8 bytes are tested for formatting, overlong encoding, surrogates and value ranges
+        /// - Invalid escaped sequence are copied to output as is
+        /// - It doesn't check if the string contains query
+        /// </summary>
+        public static int DecodeInPlace(char[] buffer)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (GetNextEncoded(0, buffer) == buffer.Length)
+            {
+                return buffer.Length;
+            }
+
+            return DecodeCore(buffer, buffer);
+        }
+
+        /// <summary>
+        /// Unescape a url path string.
+        /// 
+        /// - Everything is unescaped except %2F ('\')
+        /// - UTF8 bytes are tested for formatting, overlong encoding, surrogates and value ranges
+        /// - Invalid escaped sequence are copied to output as is
+        /// - It doesn't check if the string contains query
+        /// </summary>
+        public static string Decode(string original)
+        {
+            if (original == null)
+            {
+                throw new ArgumentNullException(nameof(original));
+            }
+
+            if (original.IndexOf('%') == -1)
+            {
+                return original;
+            }
+
+            var buffer = original.ToCharArray();
+
+            // decode in place
+            var len = DecodeCore(buffer, buffer);
+            return new string(buffer, 0, len);
+        }
+
+        /// <summary>
+        /// Unescape a url path string.
+        /// 
+        /// - The result is saved in place
+        /// </summary>
+        private static int DecodeCore(char[] source, char[] destination)
```

Comment:
This is only ever called with source == destination. Do you have a reason to maintain the difference?